test/blockcrawler/evm/test_rpc.py

```diff
@@ -730,6 +731,137 @@ async def test_get_block_number_raises_exception_for_rpc_error(self):
                 raise e
 
 
+class GetBlockByTimestampTestCase(BaseRPCClientTestCase):
+    async def asyncSetUp(self) -> None:
+        await super().asyncSetUp()
+
+        self._blocks_dataset: Dict[HexInt, HexInt] = {}
+        self._blocks_count = 1000
+
+        self._min_timestamp: Union[HexInt, None] = None
+        self._min_block_number: Union[HexInt, None] = None
+        self._max_block_number: Union[HexInt, None] = None
+
+        self._pow_min_timestamp = 1438269988
+        self._pow_min_block_number = 1
+        self._pow_max_block_number = self._blocks_count - 1
+
+        self._pos_min_timestamp = 1663224179
+        self._pos_min_block_number = 15537394
+        self._pos_max_block_number = self._pos_min_block_number + self._blocks_count - 1
+
+        self._rpc_client.get_block_number = AsyncMock()  # type: ignore
+        self._rpc_client.get_block_number.side_effect = lambda: self._max_block_number
+
+        self._rpc_client.get_block = AsyncMock()  # type: ignore
+        self._rpc_client.get_block.side_effect = lambda x: MagicMock(
+            EvmBlock, number=x, timestamp=self._blocks_dataset[x]
+        )
+
+    def _patch_with_proof_of_work_dataset(self):
+        self._min_timestamp = HexInt(self._pow_min_timestamp)
+        self._min_block_number = HexInt(self._pow_min_block_number)
+        self._max_block_number = HexInt(self._pow_max_block_number)
+
+        self._blocks_dataset.clear()
+        timestamp = self._min_timestamp
+        for block_number in range(
+            self._min_block_number.int_value, self._max_block_number.int_value + 1
+        ):
+            self._blocks_dataset[HexInt(block_number)] = timestamp
+            timestamp += random.randint(1, 50)
+
+    def _patch_with_proof_of_stake_dataset(self):
+        self._min_timestamp = HexInt(self._pos_min_timestamp)
+        self._min_block_number = HexInt(self._pos_min_block_number)
+        self._max_block_number = HexInt(self._pos_max_block_number)
+
+        self._blocks_dataset.clear()
+        timestamp = self._min_timestamp
+        for block_number in range(
+            self._min_block_number.int_value, self._max_block_number.int_value + 1
+        ):
+            self._blocks_dataset[HexInt(block_number)] = timestamp
+            # use weighted values to mimic actual timestamp differences
+            timestamp += random.choices([12, 24], weights=[9, 1])[0]
```

Comment:
Please split this into two test cases, one for proof of work and another for proof of stake. This will allow for the patching to be in the setup where it really belongs.

Reply:
Addressed.